### DISCUSSION TITLE
FlowState in Javascript

### DISCUSSION THREAD

**Discussion by Sim4n6** [github/codeql#13072] [Q&A] [answered]
Is there a similar way to use FlowState in Javascript, please ?

**Comment by aeisenberg** [comment, marked as answer]
In the JavaScript API, these are called _flow labels_. See https://codeql.github.com/docs/codeql-language-guides/using-flow-labels-for-precise-data-flow-analysis/ for a tutorial.

**Comment by Sim4n6** [comment]
Would you believe me if I tell you I was not able to set up a basic isAdditionalTaintStep with Flow labels, please?

**Reply by aeisenberg** [reply]
As a matter of fact...I _would_ believe you. 😄 

Can you please describe what you have tried so far and what exactly you are trying to achieve?

**Comment by Sim4n6** [comment]
:-D

I'm working on the Unicode Bypass Validation query version JS. 

I am using the `isAdditionalTaintStep()` as follows in a `TaintTracking::Configuration`.

```ql
 override predicate isAdditionalTaintStep(
    DataFlow::Node nodeFrom, DataFlow::FlowLabel state, DataFlow::Node nodeTo,
    DataFlow::FlowLabel state
  ) {}
```

I get this error:

```
This predicate does not override another predicate.
```

why ? maybe it is set to final on the upper class ?

**Reply by aeisenberg** [reply]
That predicate does not exist on the super class. I think @atorralba has already answered your question here:

https://github.com/github/codeql/discussions/13089#discussioncomment-5858724

His suggestion is to use `isAdditionalFlowStep(DataFlow::Node src, DataFlow::Node trg, DataFlow::FlowLabel inlbl, DataFlow::FlowLabel outlbl)` instead and specify custom labels.

**Reply by Sim4n6** [reply]
You mean 

```ql
  override predicate isAdditionalFlowStep(
    DataFlow::Node nodeFrom, DataFlow::Node nodeTo, DataFlow::FlowLabel stateFrom,
    DataFlow::FlowLabel stateTo
  ) {
    isAdditionalTaintStep(nodeFrom, nodeTo) and
    stateFrom instanceof PreValidation and
    stateTo instanceof PreValidation
  }

  predicate isAdditionalTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
//
}
```

**Reply by aeisenberg** [reply]
I think so.

**Reply by Sim4n6** [reply]
it is the warning that I was missing without a second override

```ql
  override predicate isAdditionalFlowStep(
    DataFlow::Node nodeFrom, DataFlow::Node nodeTo, DataFlow::FlowLabel stateFrom,
    DataFlow::FlowLabel stateTo
  ) {
    isAdditionalTaintStep(nodeFrom, nodeTo) and
    stateFrom instanceof PreValidation and
    stateTo instanceof PreValidation
  }

  override predicate isAdditionalTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
//
}
```

now it should work , thanks man !
